Added keyboard and mouse input remapping, mouse movement to joystick logic, GUI and more

kalaposfos13 opened the pull request:
An updated overview of the PR as of 2 months later, since so much has changed that not a lot remains from the original version:

The current state of input handling
-
- Most controllers work out of the box, with Xbox controllers' back button emulating a touchpad press as well.
- However, these are the default bindings, and they are unchangeable.
- Keyboard inputs are technically supported, but only as single buttons, you can't rebind them and the configuration is basically unplayabe.
- You can't use the mouse at all.

I decided that since more and more games are finally becoming more and more playable, an update to this system is really needed, as a core part of PC gaming is the ability to rebind controls however you want, and a lot of people prefer using KBM over controller.

An overview of the changes
-
- I added the ability to rebind your mouse, keyboard and controller inputs, with up to 3 inputs per output.
- You can create a custom configuration for every game in your library.
- I added a GUI for this (accessed by the previously unused Controller button) where you can edit your bindings straight from the launcher. You can even reload the configuration mid-game, giving you the ability to change keybinds without restarting the emulator.
- I added an implementation of mouse-to-joystick emulation, making controlling the camera in games like Bloodborne with your mouse movement possible.
- A more detailed user manual can be found by clicking the Help button in the aforementioned config editor.

A technical overview of the control flow of the new input handling system
-
0. Upon launching a game, the config is read in from the correct file. Next, the main SDL event handling loop is started. A separate timer is also created for mouse polling, which runs on a completely different system than everything else.
1. The event handling loop waits for an event. If one is found, and it is a controller, keyboard or mouse (button or wheel) event, the main part of input handling starts.
2. Based on the event, a custom input ID is calculated (these are just the SDL ID-s but with an offset based on which device it is, as all of them start at 0). For analog inputs, the distance is also stored.
3. Next, we add or remove this ID based on the event type to the list of currently active inputs (for example, we add it if it's a keydown, but remove it if it's a keyup event). If that list changed (holding down a keyboard button will generate multiple keydown events, for example, so not every event changes the list), we recalculate the state of the emulated controller, otherwise, we finish the handling of this event here.
4. If the list changed, the following things happen: First, we reset the "new" state of the controller (from now on, this refers to the emulated one). It has two states: the actual and the "new", and the later is used by the algorithm, and when it finishes, changes the actual state to the newly calculated state.
5. We then iterate over the list of all bindings. This is a sorted list, with the bindings whose inputs have more keys being higher priority than those with less. (example: [a, b] > [c] and [a, b] < [c, d, e]). For every binding, we check if all of its input ID-s are in the list of active ID-s, and if so, we flag them as "used" (The list actually stores ID - bool pairs, and the flag is reset to false every new input frame). If an input's all ID-s are found, but all of them are "used", that means the input is overridden by another one (example: [w, lalt] overrides [w], so pressing [w, lalt] doesn't activate both). This is why the bindings being sorted matters.
6. If a binding's input is determined to be valid and not overridden, we then update the corresponding output's "new" state. If it's a button, we set its "button pressed" flag to true, meaning that if any of the inputs bound to this is active, it will be pressed, and only becomes unpressed when all of those are. If it's an axis, we add the offset that is stored in the binding to it (That is either +-127 if it's a button input, or the value we read in with the input ID, and if one of the inputs is analog (you can bind an axis and a button to an axis, and that makes that axis only activate if the button is help as well)) we pass the value from that. This is what makes WASD feel much better, as now holding A and D now centers the joystick, since one adds 127 and the other adds -127, canceling each other out.
7. The actual controller state is updated from the new controller state.

This explanation leaves things like halfmode_joystick and key_toggle handling out, but it is already quite complicated without those.

kalaposfos13 committed:
added support for loading keyboard config from file

kalaposfos13 committed:
final minor update before pull request

kalaposfos13 committed:
fix messing up the merge

kalaposfos13 committed:
fix waitEvent to correctly handle mouse inputs

kalaposfos13 committed:
add license

kalaposfos13 committed:
Applied coding style fixes

kalaposfos13 committed:
clang-format fucked up the .ini file

kalaposfos13 committed:
Merge branch 'shadps4-emu:main' into main

kalaposfos13 committed:
actually fix clang changing ini syntax
use relative path for the ini file

kalaposfos13 committed:
remove big commented out code blocks,
and fixed platform-dependent code

kalaposfos13 committed:
fix windows hating me

kalaposfos13 committed:
added mouse config option

kalaposfos13 committed:
added toggle for mouse movement input (f7)

kalaposfos13 committed:
fix license and style

kalaposfos13 committed:
add numpad support i accidentally left out

kalaposfos13 committed:
added support for mouse wheel (to buttons only)

kalaposfos13 committed:
if keyboard config doesn't exist, autogenerate it

kalaposfos13 committed:
added keybinds for "walk mode"

kalaposfos13 committed:
Mouse movement input is now off by default

kalaposfos13 committed:
Merge branch 'shadps4-emu:main' into main

kalaposfos13 committed:
code cleanup and misc fixes

kalaposfos13 committed:
delete config file since it is now autogenerated

kalaposfos13 committed:
F6 = F7 + F9

kalaposfos13 committed:
Merge branch 'shadps4-emu:main' into main

kalaposfos13 committed:
added better mouse handling with config options

kalaposfos13 committed:
Added capslock support

kalaposfos13 committed:
fix clang-format

kalaposfos13 committed:
Added support for mod key toggle key

kalaposfos13 committed:
F6 and F7 are removed, F9 captures and enables the mouse

kalaposfos13 committed:
Merge remote-tracking branch 'upstream/main'

kalaposfos13 committed:
Encapsulated globals and new classes in a new namespace

kalaposfos13 committed:
Added mouse side button support

kalaposfos13 committed:
Merge remote-tracking branch 'upstream/main'

kalaposfos13 committed:
Added per-game config

kalaposfos13 committed:
relocated input parser to the new namespace

kalaposfos13 committed:
changed parser parameters to make it possible to use it from the gui

kalaposfos13 committed:
added home, end, pgup and pgdown

kalaposfos13 committed:
Merge branch 'upstreamMain'

kalaposfos13 committed:
Resolved merge conflict and refactored code

kalaposfos13 committed:
Updated default keybindings

kalaposfos13 committed:
Merge remote-tracking branch 'upstream/main' - update to 0.4.0

kalaposfos13 committed:
Changed input handling to be single-threaded

kalaposfos13 committed:
General code cleanup

kalaposfos13 committed:
Merge remote-tracking branch 'upstream/main'

kalaposfos13 committed:
Merge remote-tracking branch 'upstream/main'

raphaelthegreat commented on src/sdl_window.cpp:
```diff
@@ -19,7 +29,414 @@
+using Libraries::Pad::OrbisPadButtonDataOffset;
+
+// i wrapped it in a function so I can collapse it
+std::string getDefaultKeyboardConfig() {
```
This needs to be a separate header, and a constexpr std::string_view so it has no runtime overhead

raphaelthegreat commented on src/sdl_window.cpp:
```diff
@@ -19,7 +29,414 @@
+}
+
+// Button map: maps key+modifier to controller button
+std::map<KeyBinding, u32> button_map = {};
```
These dont belong in a windows class, please separate them in the input/ directory ideally

raphaelthegreat commented on src/sdl_window.cpp:
```diff
@@ -19,7 +29,414 @@
+    return state;
+}
+
+void parseInputConfig(const std::string game_id = "") {
```
Ditto
```suggestion
void parseInputConfig(std::string_view game_id = "") {
```

kalaposfos13 replied:
I left this as is, because it would have to be converted to std::string anyway

kalaposfos13 committed:
Start working on new backend

kalaposfos13 committed:
Mouse polling, CMakeLists, and basic framework

kalaposfos13 committed:
Output update handling, and reworked file creating, reading and parsing

kalaposfos13 committed:
Parsing works now

kalaposfos13 committed:
Single key button inputs work now

kalaposfos13 committed:
Axis outputs work now

kalaposfos13 committed:
Wheel works now (for me), l2/r2 handling improvements, and misc bugfixes

kalaposfos13 committed:
Downgraded prints to log_debug, and implemented input hierarchy

kalaposfos13 committed:
Implemented key toggle

kalaposfos13 committed:
Added mouse parameter parsing

kalaposfos13 committed:
clang-format

kalaposfos13 committed:
Merge remote-tracking branch 'upstream/main'

kalaposfos13 committed:
Merge branch 'new-kbm'

kalaposfos13 committed:
Fixed clang and added a const keyword for mac

kalaposfos13 committed:
Fix input hierarchy

kalaposfos13 committed:
Fixed joysick halfmodes, and possibly the last update on input hierarchy

kalaposfos13 committed:
clang-format

kalaposfos13 committed:
Rewrote the default config to reflect new changes

kalaposfos13 committed:
clang

kalaposfos13 committed:
Update code style

kalaposfos13 committed:
Updated sorting to accomodate for that one specific edge case

kalaposfos13 committed:
Fix default config and the latest bug with input hiearchy

kalaposfos13 committed:
Fix typo

kalaposfos13 committed:
Temporarily added my GUI

kalaposfos13 committed:
Update cmakelists

kalaposfos13 committed:
Possible fix for Gravity Rush

kalaposfos13 committed:
Update Help text, default config, and clang

kalaposfos13 committed:
Updated README with the new keybind info

kalaposfos13 committed:
okay so maybe the gravity rush fix might have slightly broken the joystick halfmode and key toggle

kalaposfos13 committed:
Fixed mistakenly overwriting the last opened config with the default one if the GUI is opened multiple times in a session

kalaposfos13 committed:
Updated Help descriptions and fixed mouse movement default parameters

kalaposfos13 committed:
Fix crash if the Help dialog was opened a second time
If it's closed with the top right close button instead of clicking the Help button again, a required flag wasn't reset, making the next click on Help try to close a nonexistent window and segfault

kalaposfos13 committed:
Added closing the config also closing the Help window, and fixed more segfaults due to mismatched flags

kalaposfos13 committed:
been a while since I last merged remote-tracking branch 'upstream/main'

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Initial controller support

kalaposfos13 committed:
clang and debug print cleanup

kalaposfos13 committed:
Initial axis-to-button logic

kalaposfos13 committed:
Updated Help text

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Added 'Reset to Default' button in GUI

kalaposfos13 committed:
Minor text and description updates + fixed an issue with Help text box rendering

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Fix button-to-touchpad logic and l2/r2 handling, as they are both axes and buttons
The touchpad's button state was correctly handled, so games that use that were fine, but the touchDown flag was always set to true, so games that use this flag had problems, like Gravity Rush

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Fix merge conflict

kalaposfos13 committed:
Clang

kalaposfos13 committed:
Added back back button to touchpad binding

kalaposfos13 committed:
Added touchpad button handling

kalaposfos13 committed:
Merge branch 'shadps4-emu:main' into main

kalaposfos13 committed:
Added end-of-line comments and fixed some crashes happening with the VS debugger

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Apply recent changes from kbm-only

kalaposfos13 committed:
Deadzone + initial directional axis-to-button mapping

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Added that one missing space in the README. Are you all happy now?

kalaposfos13 committed:
Merge remote-tracking branch 'origin/main'

kalaposfos13 committed:
Fixups from making everything use SDL

kalaposfos13 committed:
Revert directional joystick code and fix a memory leak

kalaposfos13 committed:
Change config directory name again to conform to project standards

kalaposfos13 committed:
Clang

kalaposfos13 committed:
Revert the old deeadzone code and properly add the new one

kalaposfos13 committed:
Clang
